text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,116 @@
+- `obs_service_get_url`, `obs_service_get_key`, `obs_service_get_username` and `obs_service_get_password` will be deprecated in favor of `obs_service_get_info`.
+
+List of types:
+```c
+#define OBS_SERVICE_INFO_SERVER_URL 0
+#define OBS_SERVICE_INFO_STREAM_KEY 2
+#define OBS_SERVICE_INFO_USERNAME 4 
+#define OBS_SERVICE_INFO_PASSWORD 6
+#define OBS_SERVICE_INFO_STREAM_ID 8
+#define OBS_SERVICE_INFO_ENCRYPT_PASSPHRASE 10
+```
+
+So if one day, we have a protocol that don't use a server URL, this scenario can be covered.
+
+### About `rtmp-services`
+
+This plugin will:
+- for `rtmp_common` only accept H264 or HEVC depending on the protocol, AAC and opus for FTL protocol.
+- for `rtmp_custom` it will depends on the auto-detected protocol.
+
+## UI
+
+### Audio encoders
+If the service/output is compatible with various audio encoders, the user should be able to choose in the UI which one.
+
+For AAC in simple mode, the will have only one AAC option.
+In advanced mode the user could choose between various AAC implementation.
+
+
+### Settings loading order
+
+Service settings need to be loaded first and output ones afterward to show only compatible encoder.
+
+### About service JSON file
+
+- A field `"protocol"` will be added for HLS/FTL services, other protocol can be detected through prefixes, further changes will be done in RFC 39.
+  - Services still mono-protocol, RTMP+RTMPS combo when the field is not set is the only exception.
+  - RTMP services with custom code and HTTP(S) URLs will have the protocol enforced to RTMP.
+- Services that ask for a protocol or only codecs that are not registered will not be shown.
+- `"output"` in `"recommended"` is no longer required (kept for backward compatibility) and `const char *(*get_output_type)(void *data)` in `obs_service_info` will be no longer used.
+    - It will be required through the JSON Schema if protocol different from RTMP(S) is set to avoid breakage of the backward compatibility.
+- Codecs field for audio and video will be added to enable services to allow some of codec compatible with the protocol.
+
+### Service and Output settings interactions
+NOTE: Since Services API is not usable by third-party for now. The scenario where there is no simple encoder preset for the service does not need to be considered for now.
+
+Changing protocol by changing the service should update the Output page with only compatible encoders.
+
+If the actual codec is not supported by the service, the codec will be changed and the user warned about the change.
+
+## What if there is various registered outputs for one protocol ?
+
+The improbable situation where a plugin register a output for an already registered protocol could happen, so managing this possibility is required.
+
+- In `obs_service_info`: a `const char *(*get_preferred_output_type)(void *data)` will replace `const char *(*get_output_type)(void *data)`.
```

Comment:
I'm not sure if we can outright replace `get_output_type`, or if we need to deprecate the current one and add the new one.

Reply:
I rewrote this part.